# PRUEBA TÉCNICA DE APRENDIZAJE AUTOMÁTICO

Científico de datos: Victor Manuel Rodriguez De La Hoz

# Librerías y lectura de datos

Importe de las librerías iniciales para la realización del trabajo.

In [1]:
import codecs
import pandas as pd
import os
import numpy as np
import matplotlib.pyplot as plt

df = pd.DataFrame()

Se hace la lectura de los archivos que se alojan en cada una de las carpetas entregadas para la prueba. 
Todos los archivos se agregan en orden a un dataframe que contiene dos columnas , una columna para el texto de cada muestra que se llama "Mensajes" y otra columna que indica cada texto de que carpeta es dándole un valor .

In [2]:
listaanexo=[]
listafactura=[]
listaanexoconteo1=[]
listafacturaconteo1=[]
listahistoria=[]
listahistoriaconteo1=[]
listaprocedimientos=[]
listaprocedimientosconteo1=[]
initial_count1 = 0
initial_count2 = 0
initial_count3 = 0
initial_count4 = 0
dir1 = "Anexos"
dir2 = "Factura"
dir3= "Historia"
dir4= "Procedimientos"
for path in os.listdir(dir1):
    if os.path.isfile(os.path.join(dir1, path)):
        initial_count1 += 1
for path in os.listdir(dir2):
    if os.path.isfile(os.path.join(dir2, path)):
        initial_count2 += 1
for path in os.listdir(dir3):
    if os.path.isfile(os.path.join(dir3, path)):
        initial_count3 += 1
for path in os.listdir(dir4):
    if os.path.isfile(os.path.join(dir4, path)):
        initial_count4 += 1
for i in range(0,(initial_count1-1)):
    if i==0:
        b=str(0)+str(0)+str(0)
    elif 0<i<10:
        b=str(0)+str(0)+str(i)
    elif 10<=i<100:
        b=str(0)+str(i)
    else:
        b=str(i)
    name='Anexos/doc_0'+b+'.txt'
    with codecs.open(name, encoding='latin-1') as f:
        mensaje = str(f.read())
        listaanexo.append(mensaje)
        listaanexoconteo1.append('anexo')
for i in range(0,(initial_count2-1)):
    if i==0:
        b=str(0)+str(0)+str(0)
    elif 0<i<10:
        b=str(0)+str(0)+str(i)
    elif 10<=i<100:
        b=str(0)+str(i)
    else:
        b=str(i)
    name1='Factura/doc_0'+b+'.txt'
    with codecs.open(name1, encoding='latin-1') as f:
        mensaje1 = str(f.read())
        listafactura.append(mensaje1)
        listafacturaconteo1.append('factura')
for i in range(0,(initial_count3-1)):
    if i==0:
        b=str(0)+str(0)+str(0)
    elif 0<i<10:
        b=str(0)+str(0)+str(i)
    elif 10<=i<100:
        b=str(0)+str(i)
    else:
        b=str(i)
    name2='Historia/doc_0'+b+'.txt'
    with codecs.open(name2, encoding='latin-1') as f:
        mensaje2 = str(f.read())
        listahistoria.append(mensaje2)
        listahistoriaconteo1.append('historia')
for i in range(0,(initial_count4-1)):
    if i==0:
        b=str(0)+str(0)+str(0)
    elif 0<i<10:
        b=str(0)+str(0)+str(i)
    elif 10<=i<100:
        b=str(0)+str(i)
    else:
        b=str(i)
    name3='Procedimientos/doc_0'+b+'.txt'
    with codecs.open(name3, encoding='latin-1') as f:
        mensaje3 = str(f.read())
        listaprocedimientos.append(mensaje3)
        listaprocedimientosconteo1.append('procedimientos')
df.insert(0, "Mensajes", listaanexo, allow_duplicates=False)
df=df.append([{"Mensajes": v} for v in listafactura], ignore_index=True)
df=df.append([{"Mensajes": e} for e in listahistoria], ignore_index=True)
df=df.append([{"Mensajes": h} for h in listaprocedimientos], ignore_index=True)
listaanexoconteo1+=listafacturaconteo1
listaanexoconteo1+=listahistoriaconteo1
listaanexoconteo1+=listaprocedimientosconteo1
df.insert(1, "Valor", listaanexoconteo1, allow_duplicates=False)

C:\Users\vicma\AppData\Local\Temp\ipykernel_7184\3921569738.py:86: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df=df.append([{"Mensajes": v} for v in listafactura], ignore_index=True)
C:\Users\vicma\AppData\Local\Temp\ipykernel_7184\3921569738.py:87: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df=df.append([{"Mensajes": e} for e in listahistoria], ignore_index=True)
C:\Users\vicma\AppData\Local\Temp\ipykernel_7184\3921569738.py:88: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df=df.append([{"Mensajes": h} for h in listaprocedimientos], ignore_index=True)


Como se puede ver en la visualización del dataframe se aprecia que los mensajes o textos de cada muestra tienen un texto desordenado un poco difícil de entender , para esto más adelante se realizará una limpieza del texto utilizando técnicas de NLP y en la lectura de los archivos se estudió el tipo de codificación con el que se apropia más el texto para que fuera más entendible cada texto.

In [8]:
df

Mensajes           Valor
0     Hee 9999-99 Ga Fo Nt D9\r\n\r\nClinica v.99\r\...           anexo
1     Historia: 999999 - Ingreso: 99 Pag 999 de 999\...           anexo
2     9/99/99 99:99 MATRIX - [REPORTE PARA VER EL DE...           anexo
3     Clinica\r\n\r\n \r\n\r\nauana COMPROBANTE DE P...           anexo
4     Historia: 999999 - Ingreso: 99 Pag 9 de 99\r\n...           anexo
...                                                 ...             ...
1673  GUSTAVO ALBERTO ARCILA GOMEZ\r\n\r\nHistoria: ...  procedimientos
1674  Historia: 99999 - Ingreso: 99 Pag 99 de 999\r\...  procedimientos
1675  Historia: 999999 - Ingreso: 99\r\n\r\nrire!\r\...  procedimientos
1676  Historia: 99999 - Ingreso: 9 Pag 99 de 99\r\n\...  procedimientos
1677  Historia: 999999 - Ingreso: 99 Pag 9de9\r\n\r\...  procedimientos

[1678 rows x 2 columns]

In [6]:
df['Valor'].value_counts().tolist()

[499, 499, 439, 241]

# Separación de datos de entrenamiento:

Para dividir las muestras en un conjunto de entrenamiento y uno de testeo se utiliza la librería sklearn la cual tiene un módulo que nos ayuda a dividir los datos los dos conjuntos necesarios.
En este caso se utiliza la columna "Mensajes" como los datos de texto y la columna "Valor" como el label que nos indica el valor de cada mensaje.
Como se puede ver , de las 1678 filas del dataframe se toman 1258 muestras como entrenamiento y 420 como datos de testeo, utilizando esta herramienta nos aseguramos de que se hayan distribuido proporcionalmente los datos dependiendo de cuantas muestras por categoría hay.


In [9]:
from sklearn.model_selection import train_test_split
train_text, test_text, train_labels, test_labels = train_test_split(df["Mensajes"], df["Valor"],stratify=df["Valor"])
print(f"Training examples: {len(train_text)}, testing examples {len(test_text)}")

Training examples: 1258, testing examples 420


# Preprocesamiento de datos:

Inicialmente se crea esta función ,la cual inicialmente va a recibir como dato de entrada una oración o en este caso el texto de cada muestra , utilizando este texto se utiliza el método split y realizando un condicional se revisan los caracteres que tiene cada conjunto de caracteres en el texto y procede a guardar en un nuevo arreglo solo los caracteres que no fueran signos de puntuación , estos datos se guardan pasados a minúsculas en este arreglo para posteriormente pasar cada conjunto de caracteres juntos sin espacio a un nuevo arreglo que guardará cada conjunto de caracteres que estén pegados como un dato nuevo.

En pocas palabras esta función separa por lo que puede ser una palabra el texto.


In [10]:
import string
punctuation = set(string.punctuation)
def tokenize(sentence):
    tokens = []
    for token in sentence.split():
        new_token = []
        for character in token:
            if character not in punctuation:
                new_token.append(character.lower())
        if new_token:
            tokens.append("".join(new_token))
    return tokens

En esta parte se puede apreciar lo que hace la función con los textos del dataframe.

In [11]:
df.head()["Mensajes"].apply(tokenize)

0    [hee, 999999, ga, fo, nt, d9, clinica, v99, le...
1    [historia, 999999, ingreso, 99, pag, 999, de, ...
2    [99999, 9999, matrix, reporte, para, ver, el, ...
3    [clinica, auana, comprobante, de, prestacion, ...
4    [historia, 999999, ingreso, 99, pag, 9, de, 99...
Name: Mensajes, dtype: object

Ahora se utiliza la función anteriormente creada junto al módulo "CountVectorizer”, el cual nos va a ayudar a crear una instancia que utilice la función anteriormente creada para limpiar texto y separar palabras, y a la vez este nuevo módulo revisa que la palabra aparezca una vez y le da un valor binario dentro de una lista de vocabulario conocido.

Después los datos de entrenamiento y de testeo se transforman utilizando esta función anteriormente creada.


In [12]:
from sklearn.feature_extraction.text import CountVectorizer
real_vectorizer1 = CountVectorizer(tokenizer = tokenize, binary=True)
train_X = real_vectorizer1.fit_transform(train_text)
test_X = real_vectorizer1.transform(test_text)

# Algoritmo de aprendizaje automático:

# Máquinas de vector de soporte

Para el algoritmo de aprendizaje automático se utilizó la librería sklearn con el algoritmo SVM(Maquinas de vector de soporte) el cual se ejecuta con el módulo "Linear SVC" el cual es el algoritmo ejecutado de forma linear. Este está basado en las máquinas de soporte clasificatorio.
Este algoritmo funciona haciendo un hiperplano que separa las observaciones o características propias de las categorías que se tratan y se dejan un umbral "C" el cual va a ser la distancia que puede estar separada una muestra del hiperplano de una categoría; por tanto mientras más grande sea el valor de C ,menos muestras van a pasar y las muestras se tendrán que acercar mucho más a la linealidad y a la perfección , mientras que a menor C más muestras van a pasar y el castigo en los pesos va a ser menor . 

Este algoritmo no se usa mucho en otras aplicaciones debido a que el sistema tendrá que tender a ser lineal y conlleva un coste computacional grande cuando hay muchos predictores.

La razón por la que se escoge este modelo es porque al ser un modelo en que cada categoría se diferencia por la ocurrencia de ciertas palabras en específico, el análisis de estas variables que podrían considerarse categórica puede acercarse de una forma más cercana a un sistema lineal y el modelo del hiperplano se acerca más a lo requerido. Por otro lado, este algoritmo tiene un costo computacional que depende mucho más de las observaciones que de los predictores y al tener pocas muestras y 4 categorías, el número de observaciones fue corto. Otro dato a tener en cuenta es que es un algoritmo que puede reducir el "overfitting" si se tiene en cuenta en su modo estándar, ósea que su margen no sea lleva a un máximo de hiperplano, esto quiere decir que el control de que las categorías sean memorizadas por el modelo es poco probable porque depende del umbral que tiene el modelo. A partir de esta decisión me documenté y vi que es una técnica usada en NLP.

También se pudo haber utilizado "random forest", "árbol de decisión simple”. No se utilizan regresiones por el tipo de salida esperada aunque en una posible mejora del algoritmo se puede predecir implementar con más características, columnas o información .


In [20]:
from sklearn.svm import LinearSVC
classifier = LinearSVC()
classifier.fit(train_X, train_labels)
LinearSVC(C=1.5, class_weight=None, dual=True, fit_intercept=True,
          intercept_scaling=1, loss='squared_hinge', max_iter=1000,
          multi_class='ovr', penalty='l2', random_state=None, tol=0.0001,
          verbose=0)


LinearSVC(C=1.5)

# Métricas y explicación:

Después de tener el modelo listo, primero que todo se utilizan las herramientas de la librería para medir las métricas usando los datos de prueba, posteriormente se probara el modelo manualmente.

Para la exactitud del modelo se obtuvo un 95% de precisión con un "C=1" entonces se incrementó y se varió la penalidad hasta llegar a una exactitud de 97.1429% con un "C=1.5" .

La exactitud en el modelo se entiende como que tan cerca los valores obtenidos estuvieron del valor real de la categoría, al estar en un 97% se entiende que la gran mayoría estuvieron cercanos , no hubo errores notables de overfitting y  se podría mejorar varían ciertos parámetros del modelo, la cantidad de muestras y la codificación de los datos para mejorar el NLP hecho.


In [21]:
from sklearn.metrics import accuracy_score
predicciones = classifier.predict(test_X)
accuracy = accuracy_score(test_labels, predicciones)
print(f"Accuracy: {accuracy:.4%}")

Accuracy: 97.1429%


Esta herramienta de la librería es un reporte de las métricas separadas entre cada categoría. Como se puede observar, aunque se tenían menos muestras en los anexos el valor de su precisión es el segundo más alto. La precisión se entiende como la menor dispersión entre valores obtenidos, por tanto se podría decir que los anexos tienen características bastante diferenciadores de las otras tres categorías y que posiblemente el a pocas muestras puede memorizar un poco, pero después de ver el valor de "recall" se puede decir que no memorizo porque el "recall" es la cantidad de aciertos sobre el número de muestras evaluadas, por tanto los anexos y las facturas acertaron mucho más que los demás. F1-score es la relación es una formula en la que se utiliza la precisión y el "recall", en este caso los valores fueron buenos, exceptuando el valor de los de historia que fueron los más bajos en recall y f1-score , lo cual puede decir que puede ser grande la posibilidad de desacierto o falso positivo o falso negativo en el conjunto de historia.



In [22]:
from sklearn.metrics import classification_report
reporte_metricas = classification_report(test_labels, predicciones)
print(reporte_metricas)

                precision    recall  f1-score   support

         anexo       0.98      1.00      0.99        60
       factura       1.00      1.00      1.00       125
      historia       0.97      0.92      0.94       110
procedimientos       0.94      0.98      0.96       125

      accuracy                           0.97       420
     macro avg       0.97      0.97      0.97       420
  weighted avg       0.97      0.97      0.97       420



La matriz nos dice por columna la cantidad de predicciones y por fila en que instancia fue real, hay que relacionar las predicciones por categoría y si fueron reales o falsas.
En esta matriz de confusión se pueden visualizar la cantidad de aciertos y desaciertos que tuvo el modelo. Se tuvo un desacierto en la primera categoría la cual fue confundida con la tercera categoría, la tercera categoría tuvo 3 desaciertos en la que se le dieron estos resultados a la cuarta categoría y la cuarta categoría tuvo 8 desaciertos donde estos se le dieron a la tercera categoría. Todos estos datos se corroboran más adelante en la métrica de "zero_one_loss".

De este análisis se puede deducir dos cosas directamente, la categoría tres guarda cierta relación con la categoría 1 y la categoría 4. Y la categoría 4 fue la más susceptible a errores por tanto es la que tendría que ser alimentada con más muestras o mirar la organización de los datos, o la forma de modelo.

In [23]:
from sklearn.metrics import confusion_matrix
matriz_confusion = confusion_matrix(test_labels, predicciones)
print(matriz_confusion)

[[ 60   0   0   0]
 [  0 125   0   0]
 [  1   0 101   8]
 [  0   0   3 122]]


Otra métrica importante en ingles se le conoce como "zero_one_loss", este en pocas palabras es el conteo de la cantidad de muestras en los datos de prueba que no acertaron de categoría. Si la muestra es correcta obtiene un cero, si es incorrecta obtiene un 1, estos números por muestra son sumados y este da el valor de la métrica. En este caso se obtiene zero_one_loss=12. Lo cual es un valor alto dentro del conjunto de pruebas.

In [17]:
from sklearn.metrics import zero_one_loss
zero_one_loss = zero_one_loss(test_labels, predicciones,normalize=False)
zero_one_loss

12

# Respuestas a preguntas de la prueba:

 ¿Qué efectos negativos puede tener el desbalance de muestras en el 
rendimiento de un clasificador, evidenció algún efecto negativo en este 
caso?

Los efectos negativos que puede tener el desbalance de muestras son primero que todo que es probable que el clasificador no tengo un nivel igual de alto de exactitud que con categorías con mucha mayor cantidad de muestras(mayoritario). Lo segundo que puede pasar es que el modelo no encuentre patrones o algún dato de análisis de la categoría y termine dando por acertado solo aquellos datos exclusivamente similares a los datos de entrenamiento. Por lo general en las métricas de evaluación se suele ver en estos casos que la categoría desbalanceada tiene un recall bajo y a la vez un f1-score bajo, tal vez puede tener una precisión medianamente alta. En este caso la categoría uno la cual fue la que tenía pocas muestras en comparación a las demás categorías no mostro un comportamiento marcado de que tuviera un desbalance porque sus métricas fueron buenas ; esto pudo deberse tal vez a que esta categoría tenía palabras bastante significativas y diferenciadoras de las otras categorías , la organización de sus datos pudo haber sido mejor y por ende el modelo entendió mejor la categoría tomando bien los pesos y perdidas con las que se probó el modelo dejándola así como una de las dos mejores categorías en el modelo con mejores métricas.

¿Cómo se podrían mitigar los efectos negativos del desbalance de 
muestras?

Los efectos negativos del desbalance de muestras se pueden mitigar de varias maneras.
Primero: En los parámetros del modelo se puede diferenciar los pesos de castigo de las categorías mayoritarias y las categorías minoritarias, o también dependiendo del modelo se pueden variar las pérdidas para mejorar este desbalance.

Segundo: Utilizar varios modelos en las que sus salidas sean analizadas y ajustadas para tener mayor porcentaje de precisión.

Tercero: Se pueden crear muestras sintéticas utilizando algoritmos de interpolación o relacionados que creen datos basados en un grupo minoritario de muestras. Lo no ideal es que esto puede alterar el orden natural de los datos por tal motivo es importante siempre conocer lo más posible el contexto de la categoría y el proyecto en general.

Cuarto: Modificando el dataset, esto quiere decir que se pueden reducir la cantidad de muestras del grupo mayoritario basado en el número de muestras del grupo minoritario. Esto también debe hacerse inteligentemente y tratando de que no sea totalmente manual ni totalmente aleatorio porque se pueden perder muestras significativas de la categoría mayoritaria.


# Pruebas manuales:

A continuación, se muestran pruebas manuales que se hicieron del modelo utilizando archivos aleatorios de las carpetas suministradas.
Como se puede ver las pruebas que se hicieron manualmente fueron acertadas.

In [24]:
prueba = [ """LUIS ALBERTO JARAMILLO FRANCO Pag 9 de 99
Historia: 999999 - Ingreso: 9 9

ZEl procedimiento realizado se deriva de un caso que indique presencia de politrauma? : |
No

Profilaxis Prequirargica

éRealiz9 prescripcion de profilaxis prequirurgica antibidtica? : Si |

cefazolina (uso de protesisi)

FIRMADO ELECTRONICAMENTE POR : CARLOS HERNANDO MORALES URIBE Identificacion : CC 99999999 Registro : 9999 Profesion
o Especialidad : CIRUGIA GENERAL Fecha : 9999-99-99 Hora : 99:99:99

I Wok,

Generado: 9999-99-99 - 99:99:99 por: MARIA JULIANA SOTO ARIAS


"""

]

In [25]:
prueba_X = real_vectorizer1.transform(prueba)
predicciones1 = classifier.predict(prueba_X)

In [26]:
predicciones1

array(['procedimientos'], dtype=object)

In [27]:
prueba2=["""Historia: 999999 - Ingreso: 9

Pag 99 de 999

Paciente de 99 afios , sin evaluaciones previas por psiquiatria , en el momento en periodo de duelo de fallecimiento de su esposo hace 9

meses.
No se evidencian sintomas ansiosos , afectivos , sin cambios de patrones biologicos .
Se brinda acompafiamiento

Por el momento no requiere psicofarmacos ni manejo intrahospitalario por psiquiatria
Se cierra [C y se solicita vatoracton por psicologia para acompaftamiento

Se explica a paciente y acompaftante

Conducta y Plan
Por el momento no requiere psicofarmacos ni manejo intrahospitalario por psiquiatria
Se cierra IC y se solicita vatoracion por psicologia para acompafamiento

Se explica a paciente y acompaftante

Diagnosticos CIE-99

(9) M999 OSTEOMIELITIS DE VERTEBRA.

informacion Suministrada al Paciente y/o Familia

Por el momento no requiere psicofarmacos ni manejo intrahospitalario por psiquiatria
Se cierra [C y se solicita vatoracion por psicologia para acompaftamiento
Se explica a paciente y acompafante

CLASIFICACION TIPO DE PACIENTE

CLASIFICACION TIPO DE PACIENTE :
NO COVID

FIRMADO ELECTRONICAMENTE POR : PAOLA GUTIERREZ BRICENO Identiftcacton : CC 99999999 Registro : 9999999. Profesién o

Especialidad : SIQUIATRIA Fecha : 9999-99-99 Hora : 99:99:99

*** EVOLUCION ***

Edad

99 Afios 9 Meses 9 Dias

Fecha y Hora

9999-HOSP. PISO 9 - TORRE 9 Hab. 999

 

 

9999-99-99 99:99:99

Evolucion Diaria

Generado: 9999-99-99 - 99:99:99 por: GIOVANNA HERNANDEZ
"""]

In [28]:
prueba_X2 = real_vectorizer1.transform(prueba2)
predicciones2 = classifier.predict(prueba_X2)
predicciones2

array(['historia'], dtype=object)

In [29]:
prueba3=[""" 

ADMINISTRADORA DE LOS RECURSOS DEL SISTEMA GENERAL
DE SEGURIDAD SOCIAL EN SALUD - ADRES

Informacion de Afiliados en la Base de Datos Unica de Afiliados al Sistema de Seguridad Social en Salud

Resultados de la consulta

Informacion Basica del Afiliado :

COLUMNAS DATOS
TIPO DE IDENTIFICACION cc
NUMERO DE
IDENTIFICACION 9999999999
NOMBRES SILVIA EUGENIA
APELLIDOS MOLANO MAYA
FECHA DE NACIMIENTO apes ne
DEPARTAMENTO ANTIOQUIA
MUNICIPIO LA CEJA
Datos de afiliacién :
ESTADO ENTIOAD REGIMEN FECHA DE FECHA DE ~~ ~=TIPO DE AFILIADO
AFILIACION FINALIZACION DE
EFECTIVA AFILIACION
COOMEVA
ENTIDAD
ACTIVO Ree DE CONTRIBUTIVO 99/99/9999 —«- 99/99/9999 Ss COTIZANTE
"COOMEVA
EPS. S.A."
eengge | omsanoes | estén se | sg 9a.99

La informacion registrada en esla pagina es reflejo de lo reporlado por las Enlidades en cumplimiento de la Resolucion 9999 de 9999.

Respecto a las fechas de afiliacian contenidas en esta consulta, se aclara que la Fecha de Afiliacion Efectiva hace referencia a la fecha
en la cual inicia la afiliaci9n para el usuario, la cual fue reporiada por la EPS 9 EOC, sin importar que haya estado en el Régimen
Contributivo o en el Regimen Subsidiado en dicha entidad. Ahora bien, la Fecha de Finalizacion de Afiliacian, establece el término de la
afiliacién a la entidad de acuerdo con Ia fecha de la novedad que haya presentado la EPS o EOC. A su vez se aclara que la fecha de
99/99/9999 determina que el afiliado se encuentra vinculado con la entidad que genera la consulta.

La responsabilidad por la calidad de los datos y la informacion reportada a la Base de Dalos Unica de Afiliados - BDUA, junto con el reporle
oportuno de las novedades para actualizar la BDUA, corresponde directamente a su fuente de informacion; en este caso de las EPS, EOC y
EPS-S.

Esta informacion se debe utilizar por parte de las entidades y los prestadores de servicios de salud, como complemento al marco
legal y técnico definido y nunca como motivo para denegar la prestacion do los servicios de salud a los usuarios.

Si usted encuentra una inconsistencia en la informacién publicada en ésta pagina, por favor remitase a la EPS en la cual se encuentre
afiliado y solicite la correccién de la informacion inconsistente sobre su afiliacion, Una vez realizada esta actividad, la EPS debe remitir la
novedad correspondiente a la ADRES, conforme lo establece la normatividad vigente.

IMPRIMIR CERRAR VENTANA


 """]

In [30]:
prueba_X3 = real_vectorizer1.transform(prueba3)
predicciones3 = classifier.predict(prueba_X3)
predicciones3

array(['anexo'], dtype=object)

# Conclusiones:

Se concluye que el modelo el cual tiene cuatro categorías funciona de buena manera en su presentación inicial y puede tener mejoras a futuro.

El dato desbalanceado logro adaptarse al modelo de buena manera y se propone mejorar las muestras de la categoría 3 y 4.